File: APARCH.c
#include <R.h>
#include <math.h>

// filters

// APARCH(1,1) Model Filter
void aparch_filter(int *status, double *sigma2, double* eps, double *loglik, double *param, double *y, int *T){
  
  double logden;
  double omega, alpha, gamma, beta, mu, delta;
  int t;
  
  mu    = param[0];
  omega = param[1];
  alpha = param[2];
  gamma = param[3];
  beta  = param[4];
  delta = param[5];
  
  
  // check constraints
  if( alpha <= 0 || beta < 0 || omega<0 || alpha + beta > 1){
    *loglik = -HUGE_VAL;
    return;
  }
  
  // init
  sigma2[0]=0;
  for( t=0; t<10; ++t ){ sigma2[0] += (y[t]-mu)*(y[t]-mu); }
  sigma2[0] /= 10;
  sigma2[0] *= beta;
  eps[0] = (y[0]-mu)/ pow( sigma2[0], 1/delta );
  // loop
  *loglik = 0;
  //logden    = -0.5 *log(2*M_PI) -0.5*(1/delta)*log(sigma2[0]) -0.5*((y[0]-mu)*(y[0]-mu))/pow(sigma2[0], 1/delta);
  //*loglik   += logden;
  for( t=1 ; t<*T ; ++t ){
    sigma2[t] = omega + alpha * pow( (fabs(y[t-1] - mu) + gamma * (y[t-1] - mu)) , delta) + beta * sigma2[t-1];
    eps[t]    = (y[t]-mu)/ pow( sigma2[t], 1/delta );
    
    logden    = -0.5 *log(2*M_PI) - (1/delta)*log(sigma2[t]) -0.5*((y[t]-mu)*(y[t]-mu))/pow(sigma2[t], 2/delta);
    *loglik   += logden;
  }
}

File: aparch.R
aparch11 = function(x) {
  # Estimation of APARCH(1,1) model with Gaussian innovations
  # Step 1: Initialize Time Series Globally:
  
  Tx <<- x
  N <- length(Tx)
  
  # Step 2: Initialize Model Parameters and Bounds:
  Meanx = mean(Tx); Varx = var(Tx); S = 1e-6

  params      = c(mu = Meanx, omega = 0.1*Varx, alpha = 0.1, gam1= 0.02, beta = 0.81,delta=2)
  lowerBounds = c(mu = -10*abs(Meanx), omega = S, alpha = S, gam1=-(1-S), beta = S,delta=0.1)
  upperBounds = c(mu = 10*abs(Meanx), omega = 10*Varx, alpha = 1-S, gam1 = 1-S, beta = 1-S,delta=5)
  
  aparchLLH <- function(params) {
    
    if( any(!is.finite(params)) ){ 
      return( Inf ) 
    }  
    # Given parameters, calculate the log-likelihood of Tx
    mu = params[1]; omega = params[2]; alpha = params[3]; gam1=params[4]; beta = params[5]; delta=params[6]

    z = (Tx-mu); Mean = mean(z[1:10]**2) 
    eps1 <- (sum(abs(z)))
    eps = c(mean(z), z[-N])
    e = omega + alpha*( abs( eps ) - gam1*eps )**delta 
    sigma = Mean^(delta/2)
    sigma = rep(sigma,N)
    for(i in 2:N){
      sigma[i] = beta*sigma[i-1] + e[i]
    }
    hh = abs(sigma)**(1/delta)
    
    if(any(log(dnorm(x=z, sd = hh)) == -Inf)){
      #controls for errors in input data
      idx <- which(log(dnorm(x=z, sd = hh)) == -Inf)
      z   <- z[-idx]
      hh  <- hh[-idx]
    }
    
    llh = -sum(log(dnorm(x=z, sd = hh)))
    return(llh)
  }
  
  # Step 3: Estimate Parameters and Compute Numerically Hessian:
  fit = nlminb(start = params, objective = aparchLLH,
               lower = lowerBounds, upper = upperBounds)   #,  control = list(trace=3))
  epsilon = 0.0001 * fit$par
  npar=length(params)
  Hessian = matrix(0, ncol = npar, nrow = npar)
  for (i in 1:npar) {
    for (j in 1:npar) {
      x1 = x2 = x3 = x4  = fit$par
      x1[i] = x1[i] + epsilon[i]; x1[j] = x1[j] + epsilon[j]
      x2[i] = x2[i] + epsilon[i]; x2[j] = x2[j] - epsilon[j]
      x3[i] = x3[i] - epsilon[i]; x3[j] = x3[j] + epsilon[j]
      x4[i] = x4[i] - epsilon[i]; x4[j] = x4[j] - epsilon[j]
      Hessian[i, j] = (aparchLLH(x1)-aparchLLH(x2)-aparchLLH(x3)+aparchLLH(x4))/
        (4*epsilon[i]*epsilon[j])
    }
  }
  cat("Log likelihood at MLEs: ","\n")
  print(-aparchLLH(fit$par))
  
  # Step 6: Create and Print Summary Report:
  se.coef = sqrt(abs(diag(solve(Hessian))))
  tval = fit$par/se.coef
  matcoef = cbind(fit$par, se.coef, tval, 2*(1-pnorm(abs(tval))))
  dimnames(matcoef) = list(names(tval), c(" Estimate",
                                          " Std. Error", " t value", "Pr(>|t|)"))
  cat("\nCoefficient(s):\n")
  printCoefmat(matcoef, digits = 6, signif.stars = TRUE)
  # compute output
  est=fit$par
  mu = est[1]; omega = est[2]; alpha = est[3]; gam1=est[4]; beta = est[5]; delta = est[6]
  z=(Tx-mu); Mean = mean(abs(z)**delta)
  e = omega + alpha * ( abs( c(Mean, z[-length(z)]) ) - gam1*c(Mean,z[-length(z)]) )**delta 
  h = filter(e, beta, "r", init = Mean)
  sigma.t = abs(h)**(1/delta)
  return(list(residuals = z, volatility = sigma.t, par=est))
}
